### PR TITLE
feat: Emily reorg handling. (reorg part 4/4)

### PR DESCRIPTION
**Pull request by AshtonStephens**
## Description

Closes: #348 

Handles reorgs when the chainstate updates.

Also adds a gate to other modifying API requests that prevents state alteration when the api is reorganizing. We can likely remove this gate long term but for now I'm adding it for stability.

## Changes

Adds reorganization handling.

## Testing Information

Tested with unit tests.

## Checklist:

- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by djordon — COMMENTED** [review]
I'm not done reviewing. I just thought that I'd submit this until I get back to it. Sorry if these are silly questions.

File: emily/handler/src/api/handlers/internal.rs
Comment (AshtonStephens) [reply emily/handler/src/api/handlers/internal.rs:136]: Yes. This will get all deposits / withdrawals that could have been impacted by the reorg.

File: emily/handler/src/api/handlers/internal.rs
Comment (AshtonStephens) [reply emily/handler/src/api/handlers/internal.rs:121]: That's a good question! Will fix ~

File: emily/handler/src/api/handlers/internal.rs
Comment (djordon) [reply emily/handler/src/api/handlers/internal.rs:136]: I suppose that means all requests at a height less than the input height?

Comment (AshtonStephens) [reply emily/handler/src/api/handlers/internal.rs:136]: Less than or equal to, because the one given is being reorg so it too needs to be reorged

File: emily/handler/src/api/handlers/deposit.rs
Comment (AshtonStephens) [reply emily/handler/src/api/handlers/deposit.rs:273]: I'm going to add it in a README.md.

Since this finished the MVP features of the Emily API my next PRs are:
 - OpenAPI client (Currently putting into a PR)
 - Add Readme.md
 - Add node watcher
 - Deploy in cloud

File: emily/handler/src/common/error.rs
Comment (djordon) [inline emily/handler/src/common/error.rs:26]: typo, I think it's "canonical".

File: emily/handler/src/database/entries/deposit.rs
Comment (djordon) [inline emily/handler/src/database/entries/deposit.rs:169]: Doesn't the comment imply that we should be using a `!=` here for the block_hash comparison?

File: emily/handler/src/database/entries/deposit.rs
Comment (djordon) [inline emily/handler/src/database/entries/deposit.rs:174]: nit: shouldn't this comment say reprocessing event?

File: emily/handler/src/database/entries/deposit.rs
Comment (djordon) [inline emily/handler/src/database/entries/deposit.rs:187]: Can you add a description of why this is necessary? I don't really know what synchronize means here.

File: emily/handler/tests/integration/complex/reorg.rs
Comment (djordon) [inline emily/handler/tests/integration/complex/reorg.rs:78]: I'd remove this.

File: emily/handler/src/api/handlers/internal.rs
Comment (djordon) [reply emily/handler/src/api/handlers/internal.rs:136]: Okay, I think I see. That the function name gave me the impression that we care about heights greater than the input height. Maybe we should rename it, or clarify that "after" here means that it happened before/less than the input height.

File: emily/handler/src/api/handlers/internal.rs
Comment (AshtonStephens) [reply emily/handler/src/api/handlers/internal.rs:97]: This only occurs in the following scenario:

1. This lambda gets the api state *AND* another lambda gets the API state
2. The other lambda successfully updates the api state
3. This lambda tries to update the api state but finds the version was incremented between the moment it first got the state and it attempted to update it.

A sleep only makes sense if you're waiting for some other process to finish before making another attempt, but in this case we're just trying to update the API state and encountered a race condition such that we are no longer certain what the global state of the API is.

To handle this we redo the whole process of discovering the API state again and figuring out how to handle it. Also note that if two lambdas attempt to do this at the same time at least one always succeeds, so the number of retries for a single lambda when there are `N` concurrent lambdas will always be at most `N`.

And we plan to at most have concurrent 15 maintainers of this API via sidecars on the 15 signers, so the `20` maximum number of retries is `5` more than should be theoretically possible.

File: emily/handler/src/api/handlers/internal.rs
Comment (AshtonStephens) [reply emily/handler/src/api/handlers/internal.rs:195]: Could be, I thought debug was more appropriate. If you have an opinion I'd defer to it.

File: emily/handler/src/database/entries/deposit.rs
Comment (AshtonStephens) [reply emily/handler/src/database/entries/deposit.rs:187]: I'll write it here and also add a comment in the code.

The Deposit and Withdrawal entries both contain an internal vector of history entries in chronological order. The last entry in the history vector is the latest entry, meaning the most up-to-date data. Within this history is the most recent `last_update_height` and `status`.

DynamoDB can only be sorted and indexed by top level fields, so in order to allow the table to be searchable by `status` or ordered by `last_update_height` there needs to be a top level field for it.

This function takes the entry and then synchronizes the top level fields `status` and `last_update_height` with the history vector by taking the latest history event and updating those entry fields to match it. 

I was originally doing this directly a number of places in the code and then I realized it was better to just update the history and have a separate function that synchronizes the top level entry fields with its history.

File: emily/handler/tests/integration/complex/reorg.rs
Comment (AshtonStephens) [reply emily/handler/tests/integration/complex/reorg.rs:78]: Doh!

File: emily/handler/src/api/handlers/internal.rs
Comment (AshtonStephens) [reply emily/handler/src/api/handlers/internal.rs:160]: Similar answer as before; this failed at updating an entry but successfully completing this isn't gatekept by a different process completing.

This only occurs if, between the lines `150` and `153`, another lambda manages to update the deposit.

This should actually be impossible at this moment because at this point the api status was successfully updated and all api calls that can update the deposits and withdrawals will temporarily all fail; but this is here for redundancy.

File: emily/handler/src/database/entries/deposit.rs
Comment (AshtonStephens) [reply emily/handler/src/database/entries/deposit.rs:169]: It does indeed...

File: emily/handler/src/api/handlers/internal.rs
Comment (AshtonStephens) [reply emily/handler/src/api/handlers/internal.rs:136]: I'll change it to just be after the height like you're suggesting and then put `height - 1` into the function.

File: emily/handler/src/api/handlers/chainstate.rs
Comment (AshtonStephens) [reply emily/handler/src/api/handlers/chainstate.rs:199]: Much better!

File: emily/handler/src/api/handlers/chainstate.rs
Comment (AshtonStephens) [reply emily/handler/src/api/handlers/chainstate.rs:201]: Using the following instead so that the Ok type is automatically cast:

```rust
e @ Err(_) => e?,
```

File: emily/handler/src/api/handlers/chainstate.rs
Comment (AshtonStephens) [reply emily/handler/src/api/handlers/chainstate.rs:170]: This gets killed in the next PR anyway because I delete all the type aliases so that openapi can make a sensible spec.

File: emily/handler/src/api/handlers/internal.rs
Comment (AshtonStephens) [reply emily/handler/src/api/handlers/internal.rs:75]: Done!

File: emily/handler/src/api/handlers/internal.rs
Comment (AshtonStephens) [reply emily/handler/src/api/handlers/internal.rs:59]: Done!

File: emily/handler/src/api/handlers/internal.rs
Comment (AshtonStephens) [reply emily/handler/src/api/handlers/internal.rs:44]: Done!

File: emily/handler/src/api/handlers/internal.rs
Comment (AshtonStephens) [reply emily/handler/src/api/handlers/internal.rs:94]: I prefer internal server error because more has gone wrong than a version conflict internally here than just a failed update, even if that was the reason for the deeper internal failure.

File: emily/handler/src/database/entries/deposit.rs
Comment (AshtonStephens) [reply emily/handler/src/database/entries/deposit.rs:169]: Fixed, the error was in the comment.

File: emily/handler/src/database/entries/deposit.rs
Comment (AshtonStephens) [reply emily/handler/src/database/entries/deposit.rs:174]: Done!

**Comment by AshtonStephens** [comment]
@djordon - addressed all your comments.

File: emily/handler/src/api/handlers/internal.rs
Comment (djordon) [reply emily/handler/src/api/handlers/internal.rs:116]: Wait, that's not quite what I was suggesting. The name `get_all_deposit_entries_modified_after_height` suggest to me that it will get all deposit entries that were "confirmed"/"handled" on a block height that was greater than (or greater than or equal to) the given input block height. But I think you just said that id does not do that, but instead it gets all deposit entries that happened at an earlier block height. That to me suggests that the function is misnamed (or that I misunderstand something). So you don't need to change the usage here.

File: emily/handler/src/api/handlers/internal.rs
Comment (djordon) [reply emily/handler/src/api/handlers/internal.rs:175]: Yeah, I think it this is one of the more complex parts of the API so I think it's good to start off a little noisy. If it turns out to be too noisy we can scale it back.

File: emily/handler/src/api/handlers/internal.rs
Comment (djordon) [reply emily/handler/src/api/handlers/internal.rs]: Ahh, I see. Thank you for the detailed breakdown. So if there is one sidecar sending events to the lambda serially, we wouldn't encounter this issue at all then right?

Well in either case, I think it's worth adding this comment somewhere into our documentation.

File: emily/handler/src/api/handlers/internal.rs
Comment (djordon) [inline emily/handler/src/api/handlers/internal.rs:110]: I'm not sure if this is really better, but this can be simplified to
```rust
// [`None`] means the API state is already what we want it to be, so we
// can exit early.
if set_api_state_status(context, &new_status).await?.is_none() {
    return Ok(empty_reply);
}
```

**Review by djordon — APPROVED** [review]
I just have a few suggestions/questions but they aren't critical and this basically looks good.

**Pull request merged**
